### DISCUSSION TITLE
[SOLVED] [CadQuery + PyInstaller] - ModuleNotFoundError: No module named 'OCP'

### DISCUSSION THREAD

**Discussion by bernardogbecker** [CadQuery/cadquery#1300] [Q&A]
Hi everyone!

I'm a Mechanical Engineer working on automatizing some CADs that we use here inside WEG company. Problem is that my work works flawesly in any machine that installs Conda + Cadquery + Vscode + Python... But this is completely not pratical. I need to create an executable file for any part of my company world around be able to use it.

Error:
Traceback (most recent call last):
  File "GerarSobrebases1Chapa.py", line 6, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
  File "drawing3D.py", line 1, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
  File "cadquery\__init__.py", line 10, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
  File "cadquery\occ_impl\geom.py", line 5, in <module>
ModuleNotFoundError: No module named 'OCP'

Does someone knows how to fix it? I've tried to use pyinstaller with/out --hidden-import. I've tried to use nuitka too (and gives the same problem). I've tried to import cadquery as a folder in my project. Nothing have worked out yet. 

I really need help. Thanks for all!

**Comment by jmwright** [comment]
You could look at how pyinstaller is used for the CQ-editor builds. https://github.com/jmwright/CQ-editor/blob/master/.github/workflows/pyinstaller-builds-actions.yml

**Reply by bernardogbecker** [reply]
Thanks for answering! I don't get completely how that actions works. But in practice, if I follow manually what's written there (for windows) it should work right? 

and what the code is actually doing down here? Should I just do pyinstaller "code.py" in my case?

pyinstaller --debug all pyinstaller.spec ${{ github.event.inputs.type }}
        Copy-Item C:\Miniconda3\Library\bin\libssl-1_1-x64.dll  D:\a\CQ-editor\CQ-editor\dist\CQ-editor\
        Copy-Item C:\Miniconda3\Library\bin\libcrypto-1_1-x64.dll  D:\a\CQ-editor\CQ-editor\dist\CQ-editor\
        Copy-Item D:\a\CQ-editor\CQ-editor\pyinstaller\CQ-editor.cmd D:\a\CQ-editor\CQ-editor\dist\

**Comment by jmwright** [comment]
[This line](https://github.com/jmwright/CQ-editor/blob/master/.github/workflows/pyinstaller-builds-actions.yml#L113) is probably the most relevant in the action, and there might be things to review in the [pyinstaller spec](https://github.com/jmwright/CQ-editor/blob/master/pyinstaller.spec). Notice that I had to copy some dependency libraries manually in the Action. I've found pyinstaller difficult to get working initially.

**Reply by bernardogbecker** [reply]
Hi! I'm really trying but something is going really wrong. Always the problem is:
Traceback (most recent call last):
File "GerarSobrebases1Chapa.py", line 6, in
File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
File "drawing3D.py", line 1, in
File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
File "cadquery_init_.py", line 10, in
File "PyInstaller\loader\pyimod02_importers.py", line 352, in exec_module
File "cadquery\occ_impl\geom.py", line 5, in
ModuleNotFoundError: No module named 'OCP'

If you have a little bit of time and want to help, here is my code:
https://github.com/bernardogbecker/GerarSobrebases1Chapa

downloading all of the archives in the same folder and using conda for the libraries imported, the code should work fine. The problem is when trying to use pyinstaller. If you can make it happen, or anyone in the community, you will be helping an Intern to really possible get an offer job, since this code can generate (in the first moment) 50k year of savings.

**Reply by bernardogbecker** [reply]
Hi!
After really trying to copy and paste the .spec and running line by line, the error changed, but for:
ImportError: DLL load failed while importing OCP

**Comment by bernardogbecker** [comment]
Hi everyone. I will let it open to explain how I found a way to compile it.
I was using MatplotLib, csv, numpy and cadquery in the application.

The library that I found that compiles it correctly was: https://cx-freeze.readthedocs.io/en/latest/

After compiling it and trying to open, found a lot of small .dll not implemented in the Dist file originated. But then the process was just copy and paste those in the correct folder and voala, its working now.

I highly reccomend cx-freeze, especially for debugging, because it was the most straight forward one with what should I do to work around.

Thanks for all.
